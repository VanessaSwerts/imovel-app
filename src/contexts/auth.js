import React, { createContext, useContext, useState, useEffect } from 'react'
import api from '../services/api'
import AsyncStorage from '@react-native-community/async-storage'
import * as SplashScreen from 'expo-splash-screen'

const AuthContext = createContext({
  signed: false,
  user: {}
})

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null)

  useEffect(() => {
    async function loadStorageData() {
      await SplashScreen.preventAutoHideAsync()

      const [token, userData] = await AsyncStorage.multiGet([
        'user-token',
        'user-data'
      ])

      if (token[1] && userData[1]) {
        api.defaults.headers.Authorization = `Bearer ${token[1]}`
        setUser(JSON.parse(userData[1]))
      }

      await SplashScreen.hideAsync()
    }

    loadStorageData()
  }, [])

  const signIn = async (email, password) => {
    return new Promise((resolve, reject) => {
      const data = {
        email,
        password
      }

      api
        .post('/auth/login', data)
        .then(async res => {
          const { token, ...userData } = res.data

          api.defaults.headers.Authorization = `Bearer ${token}`

          setUser(userData)

          resolve(res)

          await AsyncStorage.setItem('user-token', token)
          await AsyncStorage.setItem('user-data', JSON.stringify(userData))
        })
        .catch(err => {
          reject(err)
          console.error(err)
        })
    })
  }

  const signOut = async () => {
    await AsyncStorage.clear()

    api.defaults.headers.Authorization = ''

    setUser(null)
  }

  return (
    <AuthContext.Provider value={{ signed: !!user, user, signIn, signOut }}>
      {children}
    </AuthContext.Provider>
  )
}

export function useAuth() {
  return useContext(AuthContext)
}
