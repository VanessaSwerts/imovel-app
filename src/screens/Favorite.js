import React, { useEffect, useState } from 'react'
import { KeyboardAvoidingView, View, StyleSheet, Text, SafeAreaView, FlatList, TouchableWithoutFeedback } from 'react-native'
import { ImovelCard } from '../components'
import { useNavigation } from '@react-navigation/native'
import AsyncStorage from '@react-native-community/async-storage'
import api from '../services/api'

import colors from '../constants/colors.json'

export default function Favorite() {
  const navigation = useNavigation()

  const [favorites, setFavorites] = useState([])

  const getFavorites = async () => {
    const token = await AsyncStorage.getItem("user-token")
    if (!token) {
      navigation.reset({
        index: 0,
        routes: [{ name: "SignIn" }],
      })
      return
    }

    const config = {
      headers: {
        "Authorization": "Bearer " + token
      }
    }

    api.get(`/user/favorites`, config)
      .then((res) => {
        setFavorites(res.data)
      })
      .catch((err) => {
        console.error(err)
      })
  }

  useEffect(() => {
    getFavorites()
  }, [])

  return (
    <KeyboardAvoidingView style={styles.container} >
      <Text numberOfLiner={2} style={styles.title} >Favoritos</Text>


      <SafeAreaView style={styles.listContainer}>
        <FlatList
          data={favorites}
          keyExtractor={item => item.id.toString()}
          showsVerticalScrollIndicator={false}
          renderItem={({ item }) => {
            return (
              <TouchableWithoutFeedback
                onPress={() => navigation.navigate('IMovelDetails', { item })}
              >
                <View>
                  <ImovelCard item={item} colorIcon={'red'} />
                </View>
              </TouchableWithoutFeedback>
            )
          }}
        />
      </SafeAreaView>

    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors["platinum"],
    paddingVertical: 10,
    paddingHorizontal: 30,
  },

  title: {
    margin: 10,
    fontSize: 28,
    fontWeight: "600",
    color: colors['yellow'],
    alignSelf: 'center',
  },

  listContainer: {
    flex: 1
  }

})