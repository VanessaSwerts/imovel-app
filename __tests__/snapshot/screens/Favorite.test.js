import React from 'react'
import { create } from 'react-test-renderer'
import * as auth from '../../../src/contexts/auth'

import Favorite from '../../../src/screens/Favorite'

jest.runAllTimers()
jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      navigate: jest.fn()
    })
  }
})

describe('Favorite snapshot test', () => {
  it('render Favorite screen correctly when signed=false', () => {
    jest.spyOn(auth, 'useAuth').mockImplementation(() => ({
      signed: false
    }))
    const tree = create(<Favorite />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('render Favorite screen correctly when signed=true', () => {
    jest.spyOn(auth, 'useAuth').mockImplementation(() => ({
      signed: true
    }))
    const tree = create(<Favorite />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
