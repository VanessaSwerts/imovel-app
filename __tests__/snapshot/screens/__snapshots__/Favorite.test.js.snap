// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite snapshot test render Favorite screen correctly when signed=false 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
      "paddingHorizontal": "10%",
      "paddingVertical": "25%",
    }
  }
  testID="empty-message"
>
  <Icon
    color="#0565FC"
    name="heart-broken"
    size={120}
  />
  <View>
    <Text
      style={
        Object {
          "color": "#2C3E50",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Você ainda não possui imóveis favoritos
    </Text>
    <Text
      style={
        Object {
          "color": "#AAADB3",
          "fontSize": 18,
          "marginTop": 15,
          "textAlign": "center",
        }
      }
    >
      Vá para tela principal para visualizar os imóveis e favoritá-los!
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#0565FC",
        "borderRadius": 24,
        "height": 40,
        "justifyContent": "center",
        "marginVertical": 10,
        "opacity": 1,
        "width": 130,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
        }
      }
    >
      Tela inicial
    </Text>
  </View>
</View>
`;

exports[`Favorite snapshot test render Favorite screen correctly when signed=true 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
      "paddingHorizontal": "10%",
      "paddingVertical": "25%",
    }
  }
  testID="empty-message"
>
  <Icon
    color="#0565FC"
    name="heart-broken"
    size={120}
  />
  <View>
    <Text
      style={
        Object {
          "color": "#2C3E50",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Você ainda não possui imóveis favoritos
    </Text>
    <Text
      style={
        Object {
          "color": "#AAADB3",
          "fontSize": 18,
          "marginTop": 15,
          "textAlign": "center",
        }
      }
    >
      Vá para tela principal para visualizar os imóveis e favoritá-los!
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#0565FC",
        "borderRadius": 24,
        "height": 40,
        "justifyContent": "center",
        "marginVertical": 10,
        "opacity": 1,
        "width": 130,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
        }
      }
    >
      Tela inicial
    </Text>
  </View>
</View>
`;
